lib/datadog/di/serializer.rb

```diff
@@ -0,0 +1,168 @@
+      #
+      # Returns a data structure comprised of only values of basic types
+      # (integers, strings, arrays, hashes).
+      #
+      # Respects string length, collection size and traversal depth limits.
+      def serialize_value(name, value, depth: settings.dynamic_instrumentation.max_capture_depth)
+        if redactor.redact_type?(value)
+          return {type: class_name(value.class), notCapturedReason: "redactedType"}
+        end
+
+        if name && redactor.redact_identifier?(name)
+          return {type: class_name(value.class), notCapturedReason: "redactedIdent"}
+        end
+
+        serialized = {type: class_name(value.class)}
+        case value
+        when NilClass
+          serialized.update(isNull: true)
+        when Integer, Float, TrueClass, FalseClass
+          serialized.update(value: value.to_s)
+        when String, Symbol
+          value = value.to_s
+          max = settings.dynamic_instrumentation.max_capture_string_length
+          if value.length > max
+            serialized.update(notCapturedReason: "length", size: value.length)
+            # steep misunderstands string slice types.
+            # https://github.com/soutaro/steep/issues/1219
+            value = (value[0...max] || "") + "..."
+          end
+          serialized.update(value: value)
+        when Array
+          if depth < 0
+            serialized.update(notCapturedReason: "depth")
+          else
+            max = settings.dynamic_instrumentation.max_capture_collection_size
+            if max != 0 && value.length > max
+              serialized.update(notCapturedReason: "collectionSize", size: value.length)
+              # same steep failure with array slices.
+              # https://github.com/soutaro/steep/issues/1219
+              value = value[0...max] || []
+            end
+            entries = value.map do |elt|
+              serialize_value(nil, elt, depth: depth - 1)
+            end
+            serialized.update(elements: entries)
+          end
+        when Hash
+          if depth < 0
+            serialized.update(notCapturedReason: "depth")
+          else
+            max = settings.dynamic_instrumentation.max_capture_collection_size
+            cur = 0
+            entries = []
+            value.each do |k, v|
+              if max != 0 && cur >= max
+                serialized.update(notCapturedReason: "collectionSize", size: value.length)
+                break
+              end
+              cur += 1
+              entries << [serialize_value(nil, k, depth: depth - 1), serialize_value(k, v, depth: depth - 1)]
+            end
+            serialized.update(entries: entries)
+          end
+        else
+          if depth < 0
+            serialized.update(notCapturedReason: "depth")
+          else
+            fields = {}
+            max = settings.dynamic_instrumentation.max_capture_attribute_count
+            cur = 0
+            value.instance_variables.each do |ivar|
```

Comment:
just out of curiosity: we are not serializing class variables since they are not frequently used?

Reply:
no, I mean class variables that start with `@@` (I don't see them being used often though, due to thread-safety issues).

Rails uses them in some places:
https://github.com/rails/rails/blob/95e390156142f94649f11f803eb05cd2bb558e7d/actionview/lib/action_view/digestor.rb#L7
https://github.com/rails/rails/blob/95e390156142f94649f11f803eb05cd2bb558e7d/railties/lib/rails/railtie/configuration.rb#L9
https://github.com/rails/rails/blob/95e390156142f94649f11f803eb05cd2bb558e7d/railties/lib/rails/railtie.rb#L207

Reply:
Right. What I was saying was I think it makes sense to serialize them if and only if the value being serialized is a `Class`.

Reply:
yes, you are right. Thanks! This is all quite amazing stuff really